### ISSUE TITLE
Suggestion: more precise inference when all elements are of same generic type

### ISSUE THREAD

**Issue by Zac-HD** [python/mypy#8392]
Consider the following cases:

```python
# A list of mixed cats and dogs is a list of animals.
class Animal: pass
class Cat(Animal): pass
class Dog(Animal): pass
pets = [Cat(), Dog()]  # note: Revealed type is 'builtins.list[Animal*]'

# A list of mixed ints and strings is a list of objects.
shallow = [1, "2"]  # note: Revealed type is 'builtins.list[builtins.object*]'

# A list of mixed lists-of-integers and lists-of-strings is also a list of objects!
nested = [[1], ["2"]]  # note: Revealed type is 'builtins.list[builtins.object*]'
```

Inferring the most precise type which is a parent of all the element types makes sense, but IMO the element type of `nested` should be inferred to be `List[object]` instead of `object`.

-------------

I'm using `List` here as a placeholder for any generic type - this is really more of an issue for things like Hypothesis strategies such as `fixed_dictionaries()`, with signature `(Mapping[K, SearchStrategy[V]],) -> SearchStrategy[Mapping[K, V]]`.

This was reported downstream in HypothesisWorks/hypothesis#2349, because `fixed_dictionaries({"num": integers(), "string": text()})` works but fails to type-check unless all values are strategies for the same type or the mapping is assigned to a type-annotated variable first.

Python 3.8 and Mypy 0.761, but it's not dependent on Python version.

**Comment by ilevkivskyi** [comment]
We can probably special case list and dict literals somehow. I don't think there exists a deterministic algorithm that would infer such types consistently.

**Comment by AlexWaygood** [comment]
Closing as a duplicate of #6968
